# Data mining HW4

author: Ali ArjomandBigdeli (Student Number: 9423008)

## Question 3: Comparing implementation of fully connected NN VS. CNN in classifying MNIST

## Fully connected NN

In [1]:
from keras.datasets import mnist
from keras.preprocessing.image import load_img, array_to_img
from keras.utils.np_utils import to_categorical
from keras.models import Sequential
from keras.layers import Dense

import numpy as np
import pandas as pd

Using TensorFlow backend.


In [2]:
(X_train, Y_train),(X_test, Y_test) = mnist.load_data()

In [3]:
print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)
print(Y_test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


### Preprocessing the image data

In [4]:
image_height, image_width = 28, 28

In [5]:
X_train = X_train.reshape(60000, image_height * image_width)
X_test = X_test.reshape(10000, image_height * image_width)
print(X_train.shape)
print(X_test.shape)

(60000, 784)
(10000, 784)


In [6]:
# rescale our data
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

In [7]:
Y_train = to_categorical(Y_train, 10)
Y_test = to_categorical(Y_test, 10)
print(Y_train.shape)
print(Y_test.shape)

(60000, 10)
(10000, 10)


### Building keras model

### Network Architecture:

![FullyConnectedNN_MNIST](img/FullyNN_MNIST.JPG)

In [8]:
model = Sequential()

In [9]:
model.add(Dense(512, activation= 'relu', input_shape=(784,)))
model.add(Dense(512, activation= 'relu'))
model.add(Dense(10, activation= 'softmax'))

### Compile the model

In [10]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [11]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 512)               401920    
_________________________________________________________________
dense_2 (Dense)              (None, 512)               262656    
_________________________________________________________________
dense_3 (Dense)              (None, 10)                5130      
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


### Train the model

In [12]:
history = model.fit(X_train, Y_train, epochs=20, validation_data=(X_test, Y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 26s 428us/step - loss: 0.1839 - accuracy: 0.9446 - val_loss: 0.1009 - val_accuracy: 0.9677
Epoch 2/20
60000/60000 [==============================] - 24s 392us/step - loss: 0.0806 - accuracy: 0.9748 - val_loss: 0.0770 - val_accuracy: 0.9759
Epoch 3/20
60000/60000 [==============================] - 23s 390us/step - loss: 0.0566 - accuracy: 0.9821 - val_loss: 0.0780 - val_accuracy: 0.9778
Epoch 4/20
60000/60000 [==============================] - 24s 399us/step - loss: 0.0432 - accuracy: 0.9858 - val_loss: 0.0802 - val_accuracy: 0.9781
Epoch 5/20
60000/60000 [==============================] - 24s 396us/step - loss: 0.0343 - accuracy: 0.9889 - val_loss: 0.0709 - val_accuracy: 0.9807
Epoch 6/20
60000/60000 [==============================] - 24s 403us/step - loss: 0.0294 - accuracy: 0.9910 - val_loss: 0.0940 - val_accuracy: 0.9771
Epoch 7/20
60000/60000 [==============================] 

### Evaluating the model

In [13]:
score = model.evaluate(X_test, Y_test)

10000/10000 [==============================] - 1s 69us/step


In [14]:
score

[0.15124783596569877, 0.9803000092506409]

## CNN

### Network Architecture:

![CNN_Arch_MNIST](img/CNN_Arch_MNIST.JPG)

In [15]:
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from keras.models import Sequential
from keras.datasets import mnist
from keras.utils.np_utils import to_categorical

### load data

In [16]:
(X_train, Y_train),(X_test, Y_test) = mnist.load_data()

In [17]:
print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)
print(Y_test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


### Pre-processing

In [18]:
num_classes = 10
epochs = 3

X_train = X_train.reshape(60000, 28, 28, 1)
X_test = X_test.reshape(10000, 28, 28, 1)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0
Y_train = to_categorical(Y_train, num_classes)
Y_test = to_categorical(Y_test, num_classes)

In [19]:
print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)
print(Y_test.shape)

(60000, 28, 28, 1)
(60000, 10)
(10000, 28, 28, 1)
(10000, 10)


### Create and Compile the model

In [20]:
cnn = Sequential()

In [21]:
cnn.add(Conv2D(32, kernel_size=(5,5), input_shape=(28,28,1), padding='same', activation='relu'))

In [22]:
cnn.add(MaxPooling2D())

In [23]:
cnn.add(Conv2D(64, kernel_size=(5,5), padding='same', activation='relu'))

In [24]:
cnn.add(MaxPooling2D())

In [25]:
cnn.add(Flatten())

In [26]:
cnn.add(Dense(1024, activation='relu'))

In [27]:
cnn.add(Dense(10, activation='softmax'))

In [28]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [29]:
print(cnn.summary())

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 28, 28, 32)        832       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 14, 14, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 14, 14, 64)        51264     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 7, 7, 64)          0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 3136)              0         
_________________________________________________________________
dense_4 (Dense)              (None, 1024)              3212288   
_________________________________________________________________
dense_5 (Dense)              (None, 10)               

### Train the model

In [30]:
history_cnn = cnn.fit(X_train, Y_train, epochs=5, verbose=1, validation_data=(X_test, Y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/5
60000/60000 [==============================] - 216s 4ms/step - loss: 0.0975 - accuracy: 0.9691 - val_loss: 0.0342 - val_accuracy: 0.9896
Epoch 2/5
60000/60000 [==============================] - 216s 4ms/step - loss: 0.0347 - accuracy: 0.9889 - val_loss: 0.0281 - val_accuracy: 0.9898
Epoch 3/5
60000/60000 [==============================] - 219s 4ms/step - loss: 0.0237 - accuracy: 0.9925 - val_loss: 0.0377 - val_accuracy: 0.9886
Epoch 4/5
60000/60000 [==============================] - 223s 4ms/step - loss: 0.0184 - accuracy: 0.9942 - val_loss: 0.0266 - val_accuracy: 0.9925
Epoch 5/5
60000/60000 [==============================] - 221s 4ms/step - loss: 0.0136 - accuracy: 0.9956 - val_loss: 0.0294 - val_accuracy: 0.9924


In [31]:
cnn.save_weights('cnn-model.h5')
# cnn.load_weights('cnn-model.h5')

In [32]:
score = cnn.evaluate(X_test, Y_test)

10000/10000 [==============================] - 9s 858us/step


In [33]:
score

[0.029352097202170854, 0.9923999905586243]

## Compare two model

we obtain 99.23
% accuracy by CNN model with 5 epochs, while we obtain 98.03% accuracy by FC model with 20 epochs. This happends because CNN by using conv and pooling layers selects better features than Fully Connected network.